File: apps/server/app/domain/seek.py
"""Seek-position math, extracted from StreamEngine.seek_to_*.

Pure functions: percent <-> seconds conversion with clamping. Callers own
the duration source.
"""

from __future__ import annotations


def seconds_from_percent(percent: float, duration_seconds: float | None) -> float:
    if not duration_seconds or duration_seconds <= 0:
        return 0.0
    clamped = min(100.0, max(0.0, float(percent)))
    return clamped / 100.0 * float(duration_seconds)


def clamp_seek_seconds(seconds: float, duration_seconds: float | None) -> float:
    value = max(0.0, float(seconds))
    if duration_seconds and duration_seconds > 0:
        value = min(value, float(duration_seconds))
    return value

File: packages/domain/src/seek.ts
/** Seek-position math. Ported from app/domain/seek.py. */

export function secondsFromPercent(percent: number, durationSeconds: number | null): number {
  if (!durationSeconds || durationSeconds <= 0) return 0;
  const clamped = Math.min(100, Math.max(0, percent));
  return (clamped / 100) * durationSeconds;
}

export function clampSeekSeconds(seconds: number, durationSeconds: number | null): number {
  let value = Math.max(0, seconds);
  if (durationSeconds && durationSeconds > 0) {
    value = Math.min(value, durationSeconds);
  }
  return value;
}
